[v.0.3] Don't expose 0-dim tensors to Variable API

gchanan opened the pull request.

gchanan committed:
[v0.3] Don't expose 0-dim tensors to Variable API.

gchanan commented:
CC @colesbury

colesbury commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -43,6 +43,8 @@ PyObject * THPVariable_Wrap(Variable var)
     Py_RETURN_NONE;
   }
 
+  THPUtils_assert(var.dim() > 0, "variable must have at least 1 dimension");
```
I think throwing an exception would be better here. We don't always check if the return value is null.

apaszke commented on test/test_autograd.py:
```diff
@@ -995,6 +995,15 @@ def test_setitem_mask(self):
+                return x.select(0, 1).sum()
+
+            gradcheck(func, [x])
+            gradgradcheck(func, [x])
```
Can you add a comment for this test? I don't know what makes a combination of select and sum special

gchanan committed:
[v.0.3] Ensure grad_inputs are not ATen scalars and address review comments.

gchanan commented:
ok, I addressed the review comments.

I also had to add code making sure the grad inputs were not scalars either -- this happens in NN but not the normal autograd functions.

colesbury committed:
Remove extra parentheses

gchanan commented:
build is trying to merge into master; I'll kick it off again when the build system is fixed.

gchanan commented:
@pytorchbot retest this please

gchanan commented:
@pytorchbot retest this please
